tests/base/src/test/java/org/wildfly/security/auth/server/TestCustomRealm.java

```diff
@@ -0,0 +1,110 @@
+            return SupportLevel.UNSUPPORTED;
+        }
+
+        // this realm will be able to verify password evidences only
+        public SupportLevel getEvidenceVerifySupport(Class<? extends Evidence> evidenceType, String algorithmName)
+                throws RealmUnavailableException {
+            return PasswordGuessEvidence.class.isAssignableFrom(evidenceType) ? SupportLevel.POSSIBLY_SUPPORTED : SupportLevel.UNSUPPORTED;
+        }
+
+        public RealmIdentity getRealmIdentity(final Principal principal) throws RealmUnavailableException {
+
+            if ("myadmin".equals(principal.getName())) { // identity "myadmin" will have password "mypassword"
+                return new RealmIdentity() {
+                    public Principal getRealmIdentityPrincipal() {
+                        return principal;
+                    }
+
+                    public SupportLevel getCredentialAcquireSupport(Class<? extends Credential> credentialType,
+                                                                    String algorithmName, AlgorithmParameterSpec parameterSpec) throws RealmUnavailableException {
+                        return SupportLevel.UNSUPPORTED;
+                    }
+
+                    public <C extends Credential> C getCredential(Class<C> credentialType) throws RealmUnavailableException {
+                        return null;
+                    }
+
+                    public SupportLevel getEvidenceVerifySupport(Class<? extends Evidence> evidenceType, String algorithmName) {
+                        return PasswordGuessEvidence.class.isAssignableFrom(evidenceType) ? SupportLevel.SUPPORTED : SupportLevel.UNSUPPORTED;
+                    }
+
+                    // evidence will be accepted if it is password "mypassword"
+                    public boolean verifyEvidence(Evidence evidence) {
+                        if (evidence instanceof PasswordGuessEvidence) {
+                            PasswordGuessEvidence guess = (PasswordGuessEvidence) evidence;
+                            try {
+                                return Arrays.equals("mypassword".toCharArray(), guess.getGuess());
+
+                            } finally {
+                                guess.destroy();
+                            }
+                        }
+                        return false;
+                    }
+
+                    public boolean exists() {
+                        return true;
+                    }
+                };
+            }
+            return RealmIdentity.NON_EXISTENT;
+        }
+
+        @Override
+        public void handleRealmEvent(RealmEvent event) {
+            if (event instanceof RealmSuccessfulAuthenticationEvent) {
+                assertEquals("10.12.14.16", ((RealmSuccessfulAuthenticationEvent) event).getAuthorizationIdentity().getRuntimeAttributes().get("Source-Address").get(0));
```

Comment:
As a sanity check, would also be good to verify that it's possible to retrieve the principal in both the successful and failed cases.

Reply:
@fjuma Added